src/box/alter.cc

```diff
@@ -4418,9 +4418,24 @@ on_replace_dd_schema(struct trigger * /* trigger */, void *event)
 
 /** Unregister the replica affected by the change. */
 static int
-on_replace_cluster_clear_id(struct trigger *trigger, void * /* event */)
-{
-	replica_clear_id((struct replica *)trigger->data);
+on_replace_cluster_clear_id(struct trigger *trigger, void *event)
+{
+	struct replica *replica = (struct replica *)trigger->data;
+	struct txn_stmt *stmt = (struct txn_stmt *)event;
+	if (replica->id == instance_id) {
+		assert(stmt->row->replica_id != 0 ||
+		       recovery_state != FINISHED_RECOVERY);
+		struct replica *origin = replica_by_id(stmt->row->replica_id);
+		if (recovery_state == FINISHED_RECOVERY && origin != NULL) {
+			assert(origin->applier != NULL);
+			diag_set(ClientError, ER_LOCAL_INSTANCE_ID_IS_READ_ONLY,
+				 (unsigned)instance_id);
+			struct diag *diag = diag_get();
+			applier_kill(origin->applier, diag_last_error(diag));
```

Comment:
I think you should kill all appliers this instance has, not only the one which delivered the cluster deletion message.
Otherwise having 3 nodes in a full mesh and deleting one of them, you'll keep another connection alive, and the deleted node will continue to receive updates.

Moreover, I conducted some tests, and it seems the node which didn't participate in deletion continues to consider the deleted node as a normal one, so this also has to be fixed. Example:
```lua
-- Instance 1.
-- Step 1.
box.cfg{listen=3301, replication={3301,3302,3303}}
-- Step 2. For all the following steps here use the instance which boots first (most likely instance 1, but not 100%).
box.schema.user.grant("guest", "replication")

-- Step 3. Wait for all nodes to connect.
box.space._cluster:delete{2}

-- Instance 2.
-- Step 2.
box.cfg{listen=3302, replication={3301,3302,3303}}

-- Step 6. Try to insert smth.
box.space._schema:replace{"asdf"}

-- Step 8. Even synchro with quorum 3 succeeds.
box.ctl.promote()
box.cfg{replication_synchro_quorum=3}
box.schema.space.create('test', {is_sync=true}):create_index('pk')
box.space.test:insert{123}

-- Instance 3.
-- Step 3.
box.cfg{listen=3303, replication={3301,3302,3303}}

-- Step 4. Find the node which got deleted (in my  case it was instance 3).
box.info.id
---
- null
...
-- Step 5. Check replication: it's not fully broken.
box.info.replication
---
- 1:
    id: 1
    uuid: 3779b774-e64c-4c83-a23a-593aa9a4a0ae
    lsn: 4
    upstream:
      peer: '3301'
      lag: 0.00023484230041504
      status: stopped
      idle: 18.927021
      message: The local instance id 2 is read-only
    name: null
    downstream:
      status: follow
      idle: 0.31105300000036
      vclock: {1: 4}
      lag: 0
  3:
    id: 3
    uuid: d72536ad-39e7-4c81-9ea9-f6688f619ad3
    lsn: 0
    upstream:
      status: follow
      idle: 0.85617600000023
      peer: '3302'
      lag: 0.0002131462097168
    name: null
    downstream:
      status: follow
      idle: 0.86647700000049
      vclock: {1: 4}
      lag: 0
...

-- Step 7. See that replication continues to work on the deleted instance:
box.space._schema:fselect{}
```

Also after deletion of `instance 3` (which had id 2), `instance 2` still sees `instance 3` as a normal cluster member, even though the `_cluster` tells otherwise:
```
-- Instance 2.
2024-06-14 10:23:11.732 [39583] main/113/applier/3303 I> RAFT: message {term: 2, state: follower} from 2
2024-06-14 10:23:11.732 [39583] main/111/applier/3301 I> RAFT: message {term: 2, state: follower} from 1
tarantool> box.cfg{replication_synchro_quorum=3}
2024-06-14 10:23:21.482 [39583] main/104/interactive/box.load_cfg I> set 'replication_synchro_quorum' configuration option to 3
---
...

tarantool> box.space._cluster:fselect{}
---
- |-
  +-----+--------------------------------------+-----+
  | id  |                 uuid                 |name |
  +-----+--------------------------------------+-----+
  |  1  |"3779b774-e64c-4c83-a23a-593aa9a4a0ae"|     |
  |  3  |"d72536ad-39e7-4c81-9ea9-f6688f619ad3"|     |
  +-----+--------------------------------------+-----+
...

```

Reply:
Thanks for the catch! 

> Otherwise having 3 nodes in a full mesh and deleting one of them, you'll keep another connection alive, and the deleted node will continue to receive updates.

AFAIC, the reason this didn't happen previously, is that the deleted replica consequently tried to apply the failing deletion from all other nodes, and hence all the appliers got stopped.

Since the deletion now succeeds, it is only applied from one node. So we indeed need to stop all the appliers explicitly. This change helps stop all of the appliers, effectively stopping replication, fixing the synchronous replication issue:

```diff
diff --git a/src/box/alter.cc b/src/box/alter.cc
index 82204347b7..b6c4db59cc 100644
--- a/src/box/alter.cc
+++ b/src/box/alter.cc
@@ -4422,16 +4422,19 @@ on_replace_cluster_clear_id(struct trigger *trigger, void *event)
 {
        struct replica *replica = (struct replica *)trigger->data;
        struct txn_stmt *stmt = (struct txn_stmt *)event;
+       (void)stmt;
        if (replica->id == instance_id) {
                assert(stmt->row->replica_id != 0 ||
                       recovery_state != FINISHED_RECOVERY);
-               struct replica *origin = replica_by_id(stmt->row->replica_id);
-               if (recovery_state == FINISHED_RECOVERY && origin != NULL) {
-                       assert(origin->applier != NULL);
+               if (recovery_state == FINISHED_RECOVERY) {
                        diag_set(ClientError, ER_LOCAL_INSTANCE_ID_IS_READ_ONLY,
                                 (unsigned)instance_id);
                        struct diag *diag = diag_get();
-                       applier_kill(origin->applier, diag_last_error(diag));
+                       replicaset_foreach(replica) {
+                               if (replica->applier != NULL)
+                                       applier_kill(replica->applier,
+                                                    diag_last_error(diag));
+                       }
                        diag_clear(diag);
                }
        }
```

> Also after deletion of instance 3 (which had id 2), instance 2 still sees instance 3 as a normal cluster member, even though the `_cluster` tells otherwise:

This is reproducible on the current master branch, so if this is an issue, I don't think we should be fixing in it the scope of this patch series.

Reply:
> AFAIC, the reason this didn't happen previously, is that the deleted replica consequently tried to apply the failing deletion from all other nodes, and hence all the appliers got stopped.

Yes, indeed.

> This is reproducible on the current master branch, i.e., I didn't introduce this behavior in this patch series, so if this is an issue, I don't think we should be fixing it in the scope of this patch series.

Are you sure it's reproducible on the current master? I don't see how you could reproduce this on master if the deleted node voluntarily drops all connections to any remote peer which sends it a deletion.

The issue with your patch applied is that instance 2 already knows that instance 3 is dropped from cluster, but still counts in an ack from instance 3. Note, dropping connections from the instance 3 side won't help with this one, generally.

AFAIU the following sequence is possible:
1. Instance 1 applies the deletion of instance 3
2. Instance 2 applies the deletion of instance 3
3. Either of them sends the deletion to instance 3
4. Instance 3 starts applying a deletion, and applies a couple extra txns after it (they all go to WAL together, so the _cluster delete trigger isn't fired yet, and the connections are alive).
5. Instance 3 acks a couple of extra txns to instances 1 and 2. And instances 1/2 count instance's 3 vote/ack in. Although they shouldn't.

Reply:
> Are you sure it's reproducible on the current master? I don't see how you could reproduce this on master if the deleted node voluntarily drops all connections to any remote peer which sends it a deletion.

Under `instance 2 still sees instance 3 as a normal cluster member, even though the _cluster tells otherwise` I meant the Raft messages:

```console
2024-06-14 10:23:11.732 [39583] main/113/applier/3303 I> RAFT: message {term: 2, state: follower} from 2
```

And yes, you can also see them on current master branch. I can check again tomorrow. Didn’t check, whether this can affect elections. AFAIC, they come from the applier on the master node.

> The issue with your patch applied is that instance 2 already knows that instance 3 is dropped from cluster, but still counts in an ack from instance 3. Note, dropping connections from the instance 3 side won't help with this one, generally.

Okay, I will look into it.

Reply:
> The issue with your patch applied is that instance 2 already knows that instance 3 is dropped from cluster, but still counts in an ack from instance 3. Note, dropping connections from the instance 3 side won't help with this one, generally.

AFAIC, we have a fundamental problem here. We start sending a transaction through the relay before the `on_commit` triggers are run, so an ack from the replica can arrive before we run the `on_commit` triggers. This means that we cannot track whether an ack arrived from a deleted replica: if we try to track this from the `on_replace` trigger, we will also ignore the deletion ack from the deleted replica, making this patch useless.

Reply:
Thanks for digging into this!
You're right, it's hard to fix this.

I think the correct behavior should be:
1. ack the transaction which contains replica deletion from `_cluster`.
2. Do not ack any further transactions.

Otherwise we might end up in a situation where master commits replica deletion (2 out of 3 nodes agree, but the replica to be deleted hasn't seen the transaction yet), then master is free to add another replica in place of the old one and perform more transactions. Our old replica might receive its own deletion from cluster together with a bunch of new transactions and ack them wrongly. This can lead to a split-brain potentially.

Maybe you can stop acking everything at the moment cluster deletion is received by the replica and send a single special ack just for the transaction containing the deletion from an on_commit trigger?

Another option is to track lsns up to which a particular node has the right to ACK.

I mean, you perform _cluster deletion of a particular replica in a transaction which ends at LSN 8. Then txn_limbo may only count acks for transactions with lsn <= 8.

The second option has a problem that you have to distinguish acks not by replica ids of their senders, but by their UUIDs.

I'm summoning @Gerold103 and @Serpentian here to discuss

Reply:
3 weeks ago we discussed this with you and @Serpentian and came to a conclusion that this is not a problem, since the deleted replica is part of the cluster up the point where its deletion is confirmed.

Reply:
> I think the correct behavior should be:
> 
> 1. ack the transaction which contains replica deletion from _cluster.
> 2. Do not ack any further transactions.

Mustn't we also ack the confirm request?

Reply:
> Otherwise we might end up in a situation where master commits replica deletion (2 out of 3 nodes agree, but the replica to be deleted hasn't seen the transaction yet), then master is free to add another replica in place of the old one and perform more transactions. Our old replica might receive its own deletion from cluster together with a bunch of new transactions and ack them wrongly. This can lead to a split-brain potentially.

It doesn't seem we may cause split brain with 3 nodes, but what about 4? 

| Phase | Node 1  | Node 2      | Node 3      |Node 4      | Node 5      |
| ------------- | ------------- | ------------- |   ------------- | ------------- |------------- |
| a| 1234 | 1234 |1234 |1234 | - |
| b| 123 | 123 |123 | 1234 | - |
| c| 1235 | 1235 |123 | 1234 | - |
| d| 1235 | 1235 |123 | 1234 | 1235 |

In (a) Node 1 is a leader. In (b) we're deleting node 4 from the cluster. The deleted node doesn't ack its own deletion. In (c) new node 5 is joining to node 1. And this is the place, I don't like: in (c) node 3 may become a leader with vote from node 4. Even if PROMOTE is synchronous (and I hope, it'll be soon), node 3 will be able to commit such promote if it comes in a batch of transaction, where node 4 is deleted. Even though node 3 won't be able to commit any subsequent transactions, it's split brain.

It seems we indeed cannot continue with confirming a bunch of transactions on the node, we're going to delete. Promote shall not pass.

> Actually, I think for this case we can simply track on the master node that an ack arrived from a deleted replica using the on_commit trigger. I.e., mark a replica as deleted in the on_commit trigger and ignore acks from it further on.

Looks valid

Reply:
This is a very good discussion 💪. I agree with the above - we can't take ACKs from a node which has been deleted from `_cluster`. But I don't know if we need a flag for it. Whether we allow this replica to keep same connection or we kill its applier (and it reconnects right back anyway), in both cases `replica` object of that connection is going to have no `id`, i.e. it will be anon. And we already ignore anon acks.

A potentially very cool important point I saw here from @CuriousGeorgiy - he said *"a transaction ack from the replica can arrive before we run the transaction's on_commit triggers"*. That indeed has been the case since we have the relay, and it has shot our legs numerous times. Just look at the limbo - what kind of squats it is doing to workaround the same problem? We have to process LSN ACKs for txns which are in the limbo but have no LSN.

I think we might want to find a way to solve this problem once and forever. For example, let the relay not send ACKs into TX thread until the latter has at least same vclock or newer as the one represented by the confirmed txns. Then we would be able to safely rely on the code that `on_commit` comes first, replication comes second. It is intuitive and simplifies many places. Or there is another way. I created a ticket here: https://github.com/tarantool/tarantool/issues/10252.

Reply:
> > I think the correct behavior should be:
> > 
> > 1. ack the transaction which contains replica deletion from _cluster.
> > 2. Do not ack any further transactions.
> 
> Mustn't we also ack the confirm request?

@CuriousGeorgiy, for what? To let the leader know we have applied the deletion from _cluster? I'm not sure. Please, elaborate on that.

Reply:
> Actually, I think for this case we can simply track on the master node that an ack arrived from a deleted replica using the `on_commit` trigger. I.e., mark a replica as deleted in the `on_commit` trigger and ignore acks from it further on.

Yes, but to make it work we also need to apply @Gerold103's suggestion: make sure `on_commit` on master always happens before any acks for this transaction (#10252).

Long story short, here's the plan:
1. on_commit trigger on master node sets `replica->id = 0`
2. acks from all replicas are processed only after the corresponding on_commit trigger (simply means acks should be processed only up to current instance vclock, but not further #10252). So that after on_commit of replica deletion further acks from it are ignored.
3. on_commit trigger on replica: stops all appliers to non-anonymous(?) nodes, and stops all relays to non-anonymous nodes.

That's it. Also please make sure that as soon as `on_commit` is executed on master or on some other replicaset node, the replica can't resubscribe to this node, and shouldn't issue the rejoin.

Reply:
Sounds right. I didn't think though that we would go and do #10252 right away, but if we can, then this is very good.

Reply:
> Yes, but to make it work we also need to apply @Gerold103's suggestion: make sure on_commit on master always happens before any acks for this transaction (https://github.com/tarantool/tarantool/issues/10252).

I had the following in mind: if the deleted replica is part of the deletion quorum, then the problem does not exist. If the replica is not part of the deletion quorum, then the `on_commit` trigger on the master node is executed earlier than the ack from the deleted replica is received, so we can simply rely on setting the node as anonymous from the master node in the `on_commit` trigger.

Reply:
After oral discussion: added a `fiber_is_cancelled` check the in applier code after `latch_lock` in `applier_apply_tx`.

AFAIC, the only thing we could test here is that the replica does not ack a transaction that arrives in one batch with a confirm, however it is quite sophisticated to set this up, so I suggest to rely on the `fiber_cancel` code mentally.

Reply:
Thanks! Looks polished, finally!